CIP-hardforkSafeguard/README.md

```diff
@@ -0,0 +1,206 @@
+---
+CIP: ?
+Title: Hardfork Safeguard via Stake Representation
+Authors: Jared Corduan <jared.corduan@iohk.io>
+Status: Draft
+Type: Standards Track
+Created: 2022-06-30
+License: CC-BY-4.0
+---
+
+## Simple Summary / Abstract
+
+This CIP replaces a manual safety check regarding the readiness of the network
+for a hardfork with an automatic check.
+
+Ever since the Shelley ledger era, block headers have included a protocol
+version indicating the maximum supported protocol version that the block
+producer is capable of supporting (see section 13, Software Updates, of the
+[Shelley ledger specification](https://hydra.iohk.io/job/Cardano/cardano-ledger/shelleyLedgerSpec/latest/download-by-type/doc-pdf/ledger-spec)).
+
+This (semantically meaningless) field provides a helpful metric for determining
+how many blocks will be produced after a hardfork,
+since nodes that have not upgraded will no longer produce blocks.
+(Nodes that have not upgraded will fail the `chainChecks` check from Figure 74
+of the Shelley ledger specification, since the major protocol version in the
+ledger state will exceed the node's max major protocol version value,
+and hence can no longer make blocks.)
+
+If most of the blocks in the recent past (e.g. the last epoch) are
+broadcasting their readiness for a hardfork,
+we know that it is safe to propose an update to the major protocol version
+which triggers a hardfork.
+
+This CIP proposes automating this process,
+and making the protocol version in the header semantically meaningful.
+The ledger state will determine the stake
+(represented as the proportion of the active stake) of all the block producers
+whose last block contained the next major protocol version.
+Moreover, a new protocol parameter `hardforkThreshold` will be used to reject any
+protocol parameter update that proposes to change the major protocol version
+but does not have enough backing stake.
```

Comment:
Can you clarify was "automating the process" means in this context?
I can see in your description how this could be used to prevent a hardfork if there is insufficient backing - but what isn't explained is how such a hardfork would be initiated. Who could make such a change proposal? Is that also automated? Is it anyone who can produce blocks (so any SPO for example)? Or just some specific entity / entities?

Reply:
Let me see if I can explain it well enough here, and if it makes sense I'll add it to the CIP.

The current manual process is: around the time of a hard fork, humans use a tool like db-sync to see which stake pools are posting blocks with the current major protocol version in the block header, and which ones have the the current major protocol version plus one. If the holders of the goverance keys decide "yea, looks good", and they are otherwise ok with the hard fork (I do not myself know all that goes into the operational side if this, and am not proposing automating anything besides this single check), then they submit update proposals. If 5 of 7 agree on the exact proposals, the change occurs on the epoch boundary.

Reply:
Yup, that makes total sense... I guess what I asked for was some clarification on what exactly is meant by "automating the process" - so in the brave new world, how would this work. Is it still one of the seven proposing, at least five of seven agreeing, but THEN (after that) there is ALSO the technical requirement proposed here?
Or does this change the first part of the process as well, i.e. can others propose that change? Who needs to approve before this technical requirement comes into play.

I am thinking (but could be totally wrong) that right now you are only focused on the negative case, i.e. this CIP _only_ deals with creating a hurdle that prevents a change and doesn't try to change any other part of the process. But I think this could be somewhat clearer in this paragraph of the explanation.

Reply:
> I am thinking (but could be totally wrong) that right now you are only focused on the negative case, i.e. this CIP only deals with creating a hurdle that prevents a change and doesn't try to change any other part of the process.

That is exactly right! This CIP does not intend to change the off-chain process in any way. It's just a final blocker to stop a hard fork that is not properly endorsed by the SPOs.

Reply:
I just realized I lied :laughing: 

> This CIP does not intend to change the off-chain process in any way.

This CIP intends to lift one very specific off-chain process to the protocol itself, namely looking through block headers to see who is signaling their readiness.

Reply:
I've tried to make this clear now.